script/update-content-versions-and-check-for-archived-slugs.ts

```diff
@@ -0,0 +1,112 @@
+import createOSWebLoader from '../src/gateways/createOSWebLoader';
+import { Redirects } from '../src/redirects/types';
+
+const { bookId, newVersion } = argv as any as {
+  bookId: string
+  newVersion: string | number;
+};
+
+const booksPath = path.resolve(__dirname, '../src/config.books.json');
+const redirectsPath = path.resolve(__dirname, '../src/redirects/');
+
+const bookLoader = makeUnifiedBookLoader(
+  createArchiveLoader(`${ARCHIVE_URL}${REACT_APP_ARCHIVE_URL}`),
+  createOSWebLoader(`${ARCHIVE_URL}${REACT_APP_OS_WEB_API_URL}`)
+);
+
+async function updateRedirections(_bookId: string, currentVersion: string, _newVersion: string) {
+  if (currentVersion === _newVersion) {
+    return 0;
+  }
+
+  const { tree: currentTree, slug: bookSlug } = await bookLoader(_bookId, currentVersion)
+    .catch((error) => {
+      // tslint:disable-next-line: no-console
+      console.log(`error while loading book ${_bookId} with defaultVersion ${currentVersion}`);
+      throw error;
+    });
+
+  const { tree: newTree } = await bookLoader(_bookId, _newVersion)
+    .catch((error) => {
+      // tslint:disable-next-line: no-console
+      console.log(`error while loading book ${_bookId} with newVersion ${_newVersion}`);
+      throw error;
+    });
+
+  const redirectsBookPath = path.resolve(redirectsPath, bookId + '.json');
+
+  try {
+    // Create file with empty array only if it does not exixts
+    fs.writeFileSync(redirectsBookPath, '[]', { encoding: 'utf8', flag: 'wx' });
+  } catch {
+    // This will throw if file exists
+  }
+
+  const redirects: Redirects = require(redirectsBookPath);
+
+  const flatCurrentTree = flattenArchiveTree(currentTree);
+
+  const findRedirect = (section: LinkedArchiveTreeNode) => (
+    { pageId, bookId: pageBookId, pathname }: Redirects[0]
+  ) => pageId === section.id && pageBookId === _bookId && pathname.split('/').pop() === section.slug;
+
+  const formatSection = (section: LinkedArchiveTreeNode) => ({
+    bookId: _bookId,
+    pageId: section.id,
+    pathname: content.getUrl({ book: { slug: bookSlug }, page: { slug: section.slug } }),
+  });
+
+  let countNewRedirections = 0;
+  for (const section of flatCurrentTree) {
+    const { slug } = findArchiveTreeNodeById(newTree, section.id) || {};
+    if (
+      (slug && slug !== section.slug)
+      && !redirects.find(findRedirect(section))
+    ) {
+      redirects.push(formatSection(section));
+      countNewRedirections++;
+    }
+  }
+
+  fs.writeFileSync(redirectsBookPath, JSON.stringify(redirects, undefined, 2) + '\n', 'utf8');
+
+  return countNewRedirections;
+}
+
+async function processBook() {
+  const { defaultVersion } = books[bookId] || {};
+
+  if (defaultVersion === newVersion.toString()) {
+    console.log(`${bookId} alredy at desired version.`); // tslint:disable-line:no-console
+    process.exit(0);
+  }
+
+  const { title, version } = await bookLoader(bookId, newVersion.toString())
```

Comment:
why are you loading this twice?

Reply:
User may provide a version that is not existing so before adding it to the `config.books.json` we check if it is correct
`bookLoader` will cache results so I think it doesn't matter that it is called twice and `updateRedirections` continues to be a standalone function

Reply:
yea but couldn't you just pass the book into updateRedirections since you already have it?

Reply:
I was thinking about this but then `updateRedirections` will require `id` of the current version and the `ArchiveBook` as the third argument which makes it less generic
I was then thinking about fetching both versions of book in the main function and passing them both but when we add a new book then we don't need to fetch the current version.
I just see that I shouldn't call `updateRedirections` if there is no `defaultVersion` for the provided book since it will not find anything. I'll push a change for this in a minute